lib/interfaces.ts

```diff
@@ -41,7 +41,38 @@ export interface GooglePubSubTransportOptions {
     nackStrategy?: NackStrategy;
 }
 
-export type GenerateSubscriptionName = (topicName: string, subscriptionName?: string) => string;
+export interface TopicNameOnly {
+    _tag: 'TopicNameOnly';
+    topicName: string;
+}
+
+export interface SubscriptionNameOnly {
+    _tag: 'SubscriptionNameOnly';
+    subscriptionName: string;
+}
+
+export interface TopicAndSubscriptionNames {
+    _tag: 'TopicAndSubscriptionNames';
+    topicName: string;
+    subscriptionName: string;
+}
+
+/**
+ * The possible configurations for producing a subscription name.
+ *
+ * @remarks
+ * This allows the writer of a custom `SubscriptionNamingStrategy` to
+ * conveniently switch on the `_tag` member to enforce handling of all
+ * possible combinations of subscription and topic names.
+ *
+ * @see BasicSubscriptionNamingStrategy
+ */
+export type SubscriptionNameDependencies =
+    | TopicNameOnly
+    | SubscriptionNameOnly
+    | TopicAndSubscriptionNames;
+
+export type GenerateSubscriptionName = (deps: SubscriptionNameDependencies) => string;
```

Comment:
Do any changes need to be made to topic strategy to keep up with this?

Reply:
The topic strategy can remain as is if we want. I don't think it would benefit from any of this since it requires a single required string argument. I'm not sure there would be any benefit to making it optional either.